### PR TITLE
Theo/fix permissions

### PR DESCRIPTION
**Pull request by tnsengimana**
Add generic permission constructs that services will re-use to create permission rules. 5 other PRs have a dependency to this PR. Summary of changes:

1. Create `permissions.py` file in `zc_common.jwt_auth` and add a `BasePermission` class and other utility functions.
2. Add `PermissionTestMixin` class in `zc_common.jwt_auth.tests.py` to provide utitlity methods for testing permission classes.
3. Add unit tests in `tests.jwt_auth`

Note: As soon as this PR is merged, I will bump zc_common version to 0.2.3. All services declare that version in their requirements.txt.

### CODE REVIEW

File: tests/remote_resource/test_requests.py [hunk 1]
```diff
@@ -182,181 +182,181 @@ def test_remote_resource_list_wrapper_multiple_relationship_links(self):
...
-                }
-            ]
-        }
-
-        response_from_make_service_request = mock.Mock()
-        response_from_make_service_request.text = response_data
-        mock_get_route_from_fk.return_value = self.endpoint
-        mock_make_service_request.return_value = response_from_make_service_request
-        self.assertRaises(Exception, get_remote_resource, self.service_name, self.resource_type, self.params)
+# @mock.patch('zc_common.remote_resource.request.requests')
```
Comment (birkholz) [inline tests/remote_resource/test_requests.py]: Just delete this. If we need it again, we can grab it from git history.

File: zc_common/jwt_auth/authentication.py [hunk 2]
```diff
@@ -18,7 +18,7 @@ def __init__(self, **kwargs):
             setattr(self, kwarg, kwargs[kwarg])
 
     def is_authenticated(self):
-        return 'user' in self.roles or 'service' in self.roles
+        return True
```
Comment (birkholz) [inline zc_common/jwt_auth/authentication.py]: Why did you change this? We had `True` before and it was an issue.

Comment (birkholz) [inline zc_common/jwt_auth/authentication.py]: If anything, it should be changed to
```
return 'guest' not in self.roles
```

Which is the one case where a request would *not* be authenticated.

File: zc_common/jwt_auth/permissions.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+from rest_framework.permissions import BasePermission
+from .utils import USER_ROLE, STAFF_ROLE, SERVICE_ROLE, ANONYMOUS_ROLE
+
+
+class IsUser(BasePermission):
+    def has_permission(self, request, view):
+        user = request.user
+        return request.user and request.user.is_authenticated() and USER_ROLE in user.roles
```
Comment (birkholz) [inline zc_common/jwt_auth/permissions.py]: You have `user` as a shortcut to `request.user`, and yet you didn't use it for all instances of `request.user`?

File: zc_common/jwt_auth/permissions.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+class IsService(BasePermission):
+    def has_permission(self, request, view):
+        user = request.user
+        return request.user and request.user.is_authenticated() and SERVICE_ROLE in user.roles
+
+
+class IsStaffOrService(IsStaff, IsService):
+    def has_permission(self, request, view):
+        return IsStaff.has_permission(self, request, view) or IsService.has_permission(self, request, view)
```
Comment (birkholz) [inline zc_common/jwt_auth/permissions.py]: Both `IsStaff` and `IsService` repeat the same intro checks `request.user and request.user.is_authenticated()`, which is the majority of the permission. By calling both, you're saving maybe 1 line of redundant code at the expense of more computation time. That's a bad trade.

Also, calling the other class methods in this manner is hacky as it discards the MRO and should be avoided.

File: zc_common/jwt_auth/authentication.py (hunk 2 above)
Comment (tnsengimana) [inline zc_common/jwt_auth/authentication.py]: @birkholz I changed the line because that is in fact what permissions are doing (checking roles). Also, I am returning `True` because all requests must be authenticated by 'JWTAuthentication' class. Finally, I thought `guest` was an attribute in `mp-users` not a role?

Comment (birkholz) [inline zc_common/jwt_auth/authentication.py]: Sorry, not `guest`, the anonymous role. The role assigned to users who did not authenticate. "Unauthenticated" users should not have `is_authenticated()` return True.

File: zc_common/jwt_auth/permissions.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+class IsOwner(IsUser):
+    def has_object_permission(self, request, view, obj):
+        if not obj.user:
+            return False
+        return str(request.user.id) == str(obj.user)
+
+
+class IsAnonymous(BasePermission):
+    def has_permission(self, request, view):
+        user = request.user
```
Comment (birkholz) [inline zc_common/jwt_auth/permissions.py]: In the context of my other comment about `is_authenticated()`, this permission should be changed to just `return user.is_authenticated()` no need to check `if request.user:` either.

File: zc_common/jwt_auth/permissions.py (hunk 4 above)
Comment (tnsengimana) [inline zc_common/jwt_auth/permissions.py]: @birkholz Yes, that is very hacky because if I use `super`, both statements would be `and`ed instead of `or`ed.

Comment (tnsengimana) [inline zc_common/jwt_auth/permissions.py]: The same intro checks are independent for both IsStaff and IsService and the best alternative would be rewriting checks for both service and staff. I think that beats the purpose of composition & inheritance. There are very small microseconds spent but I think it's very trivial. Using `super` would have been ideal.

File: zc_common/jwt_auth/authentication.py (hunk 2 above)
Comment (tnsengimana) [inline zc_common/jwt_auth/authentication.py]: > The return value of is_authenticated() does not affect whether a request is authenticated by this class or not.

What do you mean? In my understanding, I would say `authentication` is the process of parsing & validating a JWT token to create a User instance. Checking whether the created user is `anonymous` or not falls within the scope of permissions.

Comment (birkholz) [inline zc_common/jwt_auth/authentication.py]: It was in response to this sentence
> Also, I am returning True because all requests must be authenticated by 'JWTAuthentication' class.

As for whether a user is anonymous or not, that has no bearing on this. The reason the `User` class has the `is_authenticated()` function in the first place is so that we maintain similarity with Django's `User` model as much as we can. Forcing true defeats the entire purpose. If it's always True, then you might as well just delete the function since it makes no difference.

File: zc_common/jwt_auth/permissions.py (hunk 4 above)
Comment (birkholz) [inline zc_common/jwt_auth/permissions.py]: I stand by my statement. It's bad practice. Please replace it.

File: zc_common/jwt_auth/authentication.py (hunk 2 above)
Comment (tnsengimana) [inline zc_common/jwt_auth/authentication.py]: Django's `.is_authenticated()` method is meant to differentiate from`models.User` from `models.Anonymous` objects. That method on each object returns either `True` or `False`, respectively ([django docs](https://docs.djangoproject.com/en/1.9/ref/contrib/auth/#django.contrib.auth.models.User.is_authenticated)). So as far as maintaining similarity is concerned, we should be returning `True`. I agree that this method/function makes no difference in our case but deleting it would not be a good idea either.

Comment (birkholz) [inline zc_common/jwt_auth/authentication.py]: Very well then. Let's have it always return `True`, but add a comment saying that checking for users with the anonymous role happens in permissions. That way no one else gets confused like I did.

File: zc_common/jwt_auth/permissions.py (hunk 4 above)
Comment (tnsengimana) [inline zc_common/jwt_auth/permissions.py]: @birkholz So this is basically what are you suggesting?

Comment (birkholz) [inline zc_common/jwt_auth/permissions.py]: How about this instead:

```python
class BaseAuthPermission(BasePermission):
  def has_permission(self, request, view):
    return request.user and request.user.is_authenticated()


class RoleCheck(BaseAuthPermission):
  passing_roles = [USER_ROLE]

  def has_permission(self, request, view):
    is_authed = super(IsStaff, self).has_permission(request, view)
    has_role = any(role in request.user.roles for role in self.passing_roles)
    return is_authed and has_role


class IsStaff(RoleCheck):
  passing_roles = [STAFF_ROLE]


class IsStaffOrService(RoleCheck):
  passing_roles = [STAFF_ROLE, SERVICE_ROLE]
```

This provides you with a framework for any combination of `OR` conditions for permissions.
If you wanted an `AND`, you'd just add multiple permissions to the relevant view.
`permission_classes = (IsStaff, IsOwner)` etc

File: zc_common/jwt_auth/permissions.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ANONYMOUS_ACTOR = 'anonymous'
+
+USER_ROLES = [USER_ACTOR]
+STAFF_ROLES = [USER_ACTOR, STAFF_ACTOR]
+SERVICE_ROLES = [SERVICE_ACTOR]
+ANONYMOUS_ROLES = [ANONYMOUS_ACTOR]
+
+
+def is_staff(request):
+    return USER_ACTOR in request.user.roles and STAFF_ACTOR in request.user.roles
```
Comment (birkholz) [inline zc_common/jwt_auth/permissions.py]: A staff user can't not have the user role also. It's unnecessary to check for it.

File: zc_common/jwt_auth/permissions.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+def is_service(request):
+    return SERVICE_ACTOR in request.user.roles
+
+
+def is_anonymous(request):
+    return ANONYMOUS_ACTOR in request.user.roles
+
+
+class BasePermission(permissions.BasePermission):
```
Comment (birkholz) [inline zc_common/jwt_auth/permissions.py]: This is a nice base class 👍

**Comment by birkholz** [comment]
Good enough to merge. :shipit:
